### PR TITLE
Adjust testing environment

### PR DESCRIPTION
**Pull request by evalott100**

### CODE REVIEW

**Comment by evalott100** [comment]
Dist is also broken on main and dev

File: src/pandablocks_ioc/ioc.py [hunk 1]
```diff
@@ -127,11 +127,11 @@ def create_softioc(host: str, record_prefix: str, screens_dir: str) -> None:
 
     try:
         dispatcher = asyncio_dispatcher.AsyncioDispatcher()
-        client = AsyncioClient(host)
         asyncio.run_coroutine_threadsafe(
             _create_softioc(client, record_prefix, dispatcher), dispatcher.loop
         ).result()
 
+        softioc.dbl()
```
Comment (coretl) [inline src/pandablocks_ioc/ioc.py:134]: I this a leftover from debugging?

File: tests/fixtures/mocked_panda.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+
+
+@pytest.fixture
+def clear_records():
+    # Remove any records created at epicsdbbuilder layer
+    ResetRecords()
+    # And at pythonSoftIoc level
+    # TODO: Remove this hack and use use whatever comes out of
+    # https://github.com/dls-controls/pythonSoftIOC/issues/56
+    RecordLookup._RecordDirectory.clear()
```
Comment (coretl) [inline tests/fixtures/mocked_panda.py:87]: Can be replaced by https://github.com/dls-controls/pythonSoftIOC/pull/114

File: tests/test_hdf_ioc.py [hunk 3]
```diff
@@ -34,38 +34,10 @@
...
-        yield chunked_read(f, 500)
-
-
-@pytest_asyncio.fixture
-def raw_dump():
-    with open(Path(__file__).parent / "raw_dump.txt", "rb") as f:
-        # Simulate largest chunked read
-        yield chunked_read(f, 200000)
-
-
```
Comment (coretl) [inline tests/test_hdf_ioc.py:68]: Are there some txt files that can be deleted from the repo now?

File: tests/test_unit_testing_structure.py
Comment (coretl) [inline tests/test_unit_testing_structure.py:1]: Testing the testing? Very meta!

File: tests/test_ioc_system.py [hunk 4]
```diff
@@ -178,216 +107,94 @@ async def test_create_softioc_update(
...
-            ]
-        )
+        assert await asyncio.wait_for(egu_queue.get(), TIMEOUT) == "s"
+        assert await asyncio.wait_for(units_queue.get(), TIMEOUT) == "s"
 
-        # Change the UNITS
+        # Change the UNITS to "min"
         assert await caput(
-            TEST_PREFIX + ":PULSE1:DELAY:UNITS", "min", wait=True, timeout=TIMEOUT
+            TEST_PREFIX + ":PULSE1:DELAY:UNITS", 0, wait=True, timeout=TIMEOUT
```
Comment (coretl) [inline tests/test_ioc_system.py:149]: Why is this not "min" still?

File: tests/test_ioc_system.py [hunk 5]
```diff
@@ -178,216 +107,94 @@ async def test_create_softioc_update(
...
-):
-    """Test that the Arm and Disarm commands are correctly sent to PandA"""
-
-    # Set the special response for the server
-    dummy_server_system.expected_message_responses.update(
-        {"*PCAP.ARM=": "OK", "*PCAP.DISARM=": "OK"}
-    )
-
-    await caput(TEST_PREFIX + ":PCAP:ARM", 1, wait=True, timeout=TIMEOUT)
-
```
Comment (coretl) [inline tests/test_ioc_system.py:213]: Do we have an equivalent of thses tests?

File: tests/test_unit_testing_structure.py
Comment (evalott100) [reply tests/test_unit_testing_structure.py:1]: I was mostly using it while trying to get the test ioc up and running, but I think it's handy to have those two tests

File: tests/test_hdf_ioc.py (hunk 3 above)
Comment (evalott100) [reply tests/test_hdf_ioc.py:68]: Yep seems like fast and slow dump aren't really used as files, neither are the `x_dump` fixtures. I'll take them all out

File: tests/test_ioc_system.py (hunk 5 above)
Comment (evalott100) [reply tests/test_ioc_system.py:213]: To me it seems like the tests are checking the `dummy_server_system` more than the ioc - asserting the dummy server responses are consumed when the message comes in from the ioc. Since we don't use the dummy server I didn't feel the need to adapt this?

I suppose I could just do the caget, caput?

Comment (coretl) [reply tests/test_ioc_system.py:213]: ok, probably don't need these tests then

Comment (AlexanderWells-diamond) [reply tests/test_ioc_system.py:213]: I believe the tests are still very important - they are confirming that updating PVs will send the correct messages to PandA to update the relevant fields/attributes there. 

Note I haven't gone through the rest of this PR yet so may be missing context.

File: tests/test_tables.py [hunk 6]
```diff
@@ -206,47 +189,30 @@ async def test_create_softioc_update_index_drvh(
...
     await caput(TEST_PREFIX + ":SEQ1:TABLE:MODE", "SUBMIT", wait=True, timeout=TIMEOUT)
 
-    # Confirm the server received the expected string
-    assert "" not in dummy_server_system.expected_message_responses
-
-    # Check the three numbers that should have updated from the REPEATS column change
-    assert "2457862145" in dummy_server_system.received
-    assert "269877249" in dummy_server_system.received
-    assert "4293918721" in dummy_server_system.received
-
```
Comment (AlexanderWells-diamond) [inline tests/test_tables.py:233]: Previously this test confirmed that the dummy server had received the expected data updates. Repeating from other discussions on this PR, it is still important that we verify the code is sending the right updates to PandA.

File: tests/test_tables.py [hunk 7]
```diff
@@ -29,8 +29,13 @@
...
 @pytest.fixture
-def table_data_dict(table_data: List[str]) -> Dict[EpicsName, RecordValue]:
-    return {EpicsName(EPICS_FORMAT_TABLE_NAME): table_data}
+def table_data_1_dict(table_data_1: List[str]) -> Dict[EpicsName, RecordValue]:
+    return {EpicsName(EPICS_FORMAT_TABLE_NAME): table_data_1}
+
+
+@pytest.fixture
+def table_data_2_dict(table_data_2: List[str]) -> Dict[EpicsName, RecordValue]:
+    return {EpicsName(EPICS_FORMAT_TABLE_NAME): table_data_2}
```
Comment (AlexanderWells-diamond) [inline tests/test_tables.py:38]: It doesn't look like this is ever used? Is there a test missing that should use it?

File: tests/test_unit_testing_structure.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+from aioca import caget
+from fixtures.mocked_panda import TEST_PREFIX
+
+
+def test_conftest_loads_fixtures_from_other_files(table_fields):
+    ...
+
+
+async def test_fake_panda_and_ioc(mocked_panda_standard_responses):
```
Comment (AlexanderWells-diamond) [inline tests/test_unit_testing_structure.py:9]: I like this. Can you just add a note/description to this test (or the whole file) to explain that it is just testing that the test framework itself works?

File: tests/fixtures/mocked_panda.py [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+                # Should never take anywhere near 10 seconds to terminate, it's just
+                # there to ensure the test doesn't hang indefinitely during cleanup
+
+    # We expect all tests to pass without warnings (or worse) logged.
+    assert (
+        len(caplog.messages) == 0
+    ), f"At least one warning/error/exception logged during test: {caplog.records}"
+
+
+def Changes_iterator_wrapper(values=None, multiline_values=None):
```
Comment (AlexanderWells-diamond) [inline tests/fixtures/mocked_panda.py:336]: Please fix the casing of this function.

File: tests/fixtures/mocked_panda.py [hunk 10]
```diff
@@ -0,0 +1,479 @@
...
+    changes = Changes(
+        values={},
+        no_value=[],
+        in_error=[],
+        multiline_values={},
+    )
+    if number_of_iterations:
+        return repeat(changes, number_of_iterations)
+
+    return repeat(changes)
```
Comment (AlexanderWells-diamond) [inline tests/fixtures/mocked_panda.py:355]: Pedantic, but isn't it unecessary to have the `if` statement? According to [the docs](https://docs.python.org/3/library/itertools.html#itertools.repeat) it looks like passing None is the same as not specifying the number of iterations at all.

File: tests/fixtures/table_data_for_tests.py
Comment (AlexanderWells-diamond) [inline tests/fixtures/table_data_for_tests.py:1]: It's a bit tautological to have `_for_tests` in the file name. It's already in a `tests` directory.

**Review by AlexanderWells-diamond — CHANGES_REQUESTED** [review]
I like the overall structural changes to the tests - it's much nicer to not have the `dummy_server`, and its associated process and network calls. It should be much easier going forward to reason about how the tests will run.

I'm requesting changes as I think important tests have been removed in this refactoring. None of the tests that checked the dummy server have received expected messages have been kept, which means we have no verification that this module is sending any of the expected messages to PandA. Given our talks about Mocks I expected to see some mocking of the `PandA-client` calls and validating that they have received the messages that we previously saw in the dummy server.

File: tests/test_ioc_system.py (hunk 5 above)
Comment (evalott100) [reply tests/test_ioc_system.py:213]: @AlexanderWells-diamond

Okay, given the new structure, I have now made a new multiprocessing queue which captures all commands passed to the panda. This will be used to convert these two unit tests

File: tests/fixtures/table_data_for_tests.py
Comment (evalott100) [reply tests/fixtures/table_data_for_tests.py:1]: There's already a `table_data` fixture or two so I didn't just want `table_data.py`, I'll name the file something else  though

File: tests/fixtures/mocked_panda.py (hunk 10 above)
Comment (evalott100) [reply tests/fixtures/mocked_panda.py:355]: Mypy error, the second argument is expecting an int, so an Optional[int] isn't allowed, I could cast it.

Comment (evalott100) [reply tests/fixtures/mocked_panda.py:355]: Think it should work with 0 being unlimited

File: tests/test_tables.py (hunk 7 above)
Comment (evalott100) [reply tests/test_tables.py:38]: I don't think so, there's a table test that needs the `table_data_1_dict`, I probably accidentaly made a second for the other table data

File: tests/test_ioc_system.py [hunk 11]
```diff
@@ -178,144 +107,44 @@ async def test_create_softioc_update(
...
-    assert "*PCAP.DISARM=" not in dummy_server_system.expected_message_responses
-
-
 def test_ensure_block_number_present():
     assert _ensure_block_number_present("ABC.DEF.GHI") == "ABC1.DEF.GHI"
     assert _ensure_block_number_present("JKL1.MNOP") == "JKL1.MNOP"
 
 
-@pytest.mark.asyncio
-async def test_create_softioc_time_panda_changes(
```
[the remark this answers is not in the document]
Comment (evalott100) [reply tests/test_ioc_system.py:227]: The contents of this test are copied in full into the next test 

``` python
async def test_create_softioc_time_epics_changes()
```

In the removed first test there was
```
        assert await asyncio.wait_for(egu_queue.get(), TIMEOUT) == "s"
        assert await asyncio.wait_for(units_queue.get(), TIMEOUT) == "s"
        assert await asyncio.wait_for(drvl_queue.get(), TIMEOUT) == 8e-09
```

But in my test environment my `drvl` ends up as `1.333333333e-10`, consistent with a conversion from `ms` to `min`, even though I'm converting from `ms` to `sec` (and the units field correctly makes this change). I'm trying to find out if this is an error in my test, or in `update_drvl()`, since my test client doesn't  ever convert  to minutes through the `MockedAsyncioClient` changes.

In the `test_create_softioc_time_epics_changes()` we convert to minutes with a caput and that does work and gives the correct `drvl`.

Thoughts on this? @AlexanderWells-diamond

[the remark this answers is not in the document]
Comment (AlexanderWells-diamond) [reply tests/test_ioc_system.py:227]: They are not quite the same test. One test checks that a PandA change is reflected into EPICS, and the other tests that an EPICS change is reflected into PandA. Observe that this deleted test has no `caput` calls - it's only watching for changes that come over the wire from the `*CHANGES?` requests. The other test uses a `caput` to trigger the opposite behaviour.

It is certainly possible to do both aspects in one test, but it seems cleaner to me to keep the separation.

As for the conversion, I'm afraid I'm finding it a little hard to follow the expected behaviour of the `standard_responses`. It looks like it creates the "DELAY" `TimeFieldInfo`, then sets its ".MIN" parameter to 1.3e-10 (the conversion from ms to minutes, but apparently without changing the units?), then changes the units to ms with an associated change in .MIN,  and then lastly changes the units to s but with no MIN change. It's hard to see what it's supposed to be.

I would suggest setting up a camonitor on the IOC, outside of the tests, and just observe what is definitely happening on the PV (and its various fields).


Also of note is that these tests did use a specialised `dummy_server_time`, rather than the generic one, as they needed slightly different data. Perhaps have another look back through that and see if there's a missing piece of data somewhere?

File: tests/test_ioc_system.py [hunk 12]
```diff
@@ -324,71 +265,87 @@ async def test_create_softioc_time_epics_changes(
...
+        command_to_key(Put(field="PCAP1.TRIG_EDGE", value="Falling"))
+        in commands_recieved_by_panda
+    )
+
+
+@pytest.mark.asyncio
+async def test_create_softioc_arm_disarm(
+    mocked_panda_standard_responses,
+):
+    """Test that the Arm and Disarm commands are correctly sent to PandA"""
```
Comment (AlexanderWells-diamond) [inline tests/test_ioc_system.py:338]: This only appears to be testing Arm, not Disarm? Ideal world they'd be two tests, but as you have to set up Arm before you can Disarm I think that's why I originally did both in one.

File: tests/test_ioc_system.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (evalott100) [reply tests/test_ioc_system.py:227]: Ahhh I figured it out, the `update_drvl` gets the `MIN` directly from the panda and uses that for the conversion

https://github.com/PandABlocks/PandABlocks-ioc/blob/9c6495ea39f93a1874524fd94020c53a56014909/src/pandablocks_ioc/ioc.py#L498

I'll just have to change the test, or fake panda responses a little.

File: tests/test_ioc_system.py (hunk 12 above)
Comment (evalott100) [reply tests/test_ioc_system.py:338]: I must have forgotten to add a disarm, doing it now.

**Comment by evalott100** [comment]
> Given our talks about Mocks I expected to see some mocking of the `PandA-client` calls and validating that they have received the messages that we previously saw in the dummy server.

I've now passed a `multiprocessing.Queue` to the `MockedAsyncioClient` so any command it receives from the softioc is put there and can be retrieved in tests.

File: tests/fixtures/mocked_panda.py [hunk 13]
```diff
@@ -0,0 +1,539 @@
...
+
+@pytest_asyncio.fixture
+def mocked_time_record_updater():
+    """An instance of _TimeRecordUpdater with MagicMocks and some default values"""
+    base_record = MagicMock()
+    base_record.name = TEST_PREFIX + ":BASE:RECORD"
+
+    # We don't have AsyncMock in Python3.7, so do it ourselves
+    client = MagicMock()
+    dummy_loop = asyncio.BaseEventLoop()
```
Comment (evalott100) [inline tests/fixtures/mocked_panda.py:63]: Without a `loop=`, python 3.10 will throw the error `DeprecationWarning: There is no current event loop`, this doesn't happen on 3.9 and 3.11 strangely.

File: pyproject.toml [hunk 14]
```diff
@@ -19,7 +19,7 @@ dependencies = [
     "h5py",
     "softioc>=4.3.0",
     "pandablocks>=0.3.1",
-    "pvi[cli]>=0.4",
+    "pvi[cli]<0.5",
```
Comment (evalott100) [inline pyproject.toml:22]: pvi requires 3.10 for v0.5, it's neccessary to pin this for consistency on `.bob` files between 3.9 and 3.10/3.11.

Maybe we should consider requiring 3.10 for pandablocks-ioc @AlexanderWells-diamond?

Comment (evalott100) [reply pyproject.toml:23]: https://github.com/PandABlocks/PandABlocks-ioc/pull/26 this will require `pvi>=0.5`

Comment (AlexanderWells-diamond) [reply pyproject.toml:23]: I don't know why PVI updated itself to Python 3.10 - it appears to be [this PR](https://github.com/epics-containers/pvi/pull/34) that did it, but there are no functional changes to the Python code that warrant dropping lower Python versions - the only changes appear to be in the `typing` module, which will not affect runtime behaviour.

Ultimately I don't really mind what version requirements we have. I'll defer to whatever @coretl and @GDYendell want to do.

File: tests/test_ioc_system.py [hunk 15]
```diff
@@ -324,71 +263,91 @@ async def test_create_softioc_time_epics_changes(
...
+):
+    """Test that updating a record causes the new value to be sent to PandA"""
+    (
+        tmp_path,
+        child_conn,
+        response_handler,
+        command_queue,
+    ) = mocked_panda_standard_responses
+
+    await asyncio.sleep(1.5)
```
Comment (AlexanderWells-diamond) [inline tests/test_ioc_system.py:322]: There's a little inconsistency regarding sleeps at the start of tests. Some tests sleep for 1 second, this one sleeps for 1.5, and others do not sleep at all. It would be good to standardise this - if all tests should be sleeping, then moving it into the fixture would be neater. 

However I'm not immediately sure that this sleep should be necessary at all - the `create_subprocess_ioc_and_responses()` function already explicitly waits for the PandA IOC to have started before signalling that it's ready. So all `caput` should be valid from the get-go.

**Review by AlexanderWells-diamond — APPROVED** [review]
Looks good now. I've just raised a couple more very minor comments, but they don't have to be resolved to merge this PR.

File: tests/test_ioc_system.py (hunk 15 above)
Comment (evalott100) [reply tests/test_ioc_system.py:322]: I've now adjusted the subprocesses so that a ready message is sent through a pip when the test is ready to start, the fixture waits for the subprocess to send this. This means we can skip wait messages!
